src/cargo/core/resolver/mod.rs

```diff
@@ -986,21 +1077,51 @@ impl<'a> Context<'a> {
         // Record what list of features is active for this package.
         if !used_features.is_empty() {
             let pkgid = candidate.package_id();
-            self.resolve_features.entry(pkgid.clone())
-                .or_insert(HashSet::new())
-                .extend(used_features);
+
+            let mut set = self.resolve_features.entry(pkgid.clone())
+                              .or_insert_with(HashSet::new);
+            for feature in used_features {
+                if !set.contains(feature) {
```

Comment:
Can microoptimize to `set.entry(feature).or_insert_with(|| feature.to_string())`.